### PR TITLE
Avoid anonymous functions to be used in arguments/callbacks

### PR DESCRIPTION
**Pull request by gedu**
Avoid anonymous callback/arguments functions

Idea: https://github.com/Expensify/App/issues/38394

Basic explantation related to allowed arrow functions:

✅ Arrow functions assignments

```jsx
const workingNamed = () => {};
```

✅ Arrow functions in objects

```jsx
const person = {getName: () => {}};
```
✅ Named functions callback

```jsx
function mappingPeople(person) {};
people.map(mappingPeople);
people.filter(function filterPerson() {});
useEffect(function handlingConnection() {}, []);
```

❌ callback arrow functions

```jsx
people.map(() => {});
people.filter(() => {});
useEffect(() => {}, []);
useEffect(function() {}, []);
```
---
1: adding a name to an anonymous function used by `useMemo`
```
const options = useMemo(
        function orderingReports() {
           // code
           return value;
        }, []);
```
2: creating an arrow function and storing it into a variable
```
const filteringForSideBar = (report) => { // code };
someValues.filter(filteringForSideBar);
```
3: using a function from an Object
```
const forEachObj = {
        callback: (report) => { //code }
}
```
named_functions

Without all the above ways we only see `(anonymous)`, and many of the functions wouldn't show up.

named_functions_anonymous

---
1: passing a function to an `useEffect`
```
useEffect(() => {
        console.log('1BEING ANONYMOUS');
        fibonacci(20);
    }, []);
```
2: passing a named function
```
useEffect(function solvingFibEleven() {
        console.log('2NotBEING ANONYMOUS');
        fibonacci(20);
    }, []);
```    
named_functions_useEffect

---
You can also search for those functions, making it easier to navigate through the performance graph

named_functions_search

### CODE REVIEW

File: eslint-plugin-expensify/avoid-anonymous-functions.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+            },
+            "CallExpression": function (node) {
+                if (node.arguments && node.arguments.some(arg => arg.type === "ArrowFunctionExpression" && !arg.id)) {
+                    context.report({
+                        node,
+                        message,
+                    });
+                }
+            },
+            "ReturnStatement > FunctionExpression, ReturnStatement > ArrowFunctionExpression": function(node) {
```
Comment (mountiny) [inline eslint-plugin-expensify/avoid-anonymous-functions.js:22]: NAB
```suggestion
            "ReturnStatement > FunctionExpression, ReturnStatement > ArrowFunctionExpression": function (node) {
```

**Review by mountiny — COMMENTED** [review]
One comment

**Comment by mountiny** [comment]
@gedu bump on the above

**Comment by gedu** [comment]
@mountiny comment applied 

> Chatting with some devs, we think would be good when it is added into Exfy, should warn, and start a migration, enforcing from C+ to avoid anonymous functions, educating devs. Now this lint rule, ignoring tests, found ±2500 cases to update

**Review by mountiny — APPROVED** [review]
@gedu I think that we should not have any coordinated migration for this, maybe as part of the performance project, you could proactively update those which are causing bottlenecks (in those files) but other than that we can keep this more to be resolved organically

**Comment by mountiny** [comment]
@gedu seems like not all commits are signed 🤔 

image

**Comment by gedu** [comment]
New Clean PR: https://github.com/Expensify/eslint-config-expensify/pull/89

**Pull request closed without merging**